### ISSUE TITLE
Incompatibility with CraftBook gate mechanic.

### ISSUE THREAD

**Issue by mibby** [IntellectualSites/FastAsyncWorldEdit-1.13#51]
Paper dev 525 (Spigot 1.13.2)
FAWE 1.13 compiled as of commit https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/commit/24fbc86cdd79b8ce13359342bc422ce1fb25a0d2
CraftBook dev 4484

CraftBook gate mechanic does not work with FAWE. The gate refuses to find/set the blocks to toggle. Downgrading to vanilla worldedit, CraftBook mechanics work fine again.

**Comment by AeSix** [comment]
Does adding CraftBook to the "allowed-plugins" section in the FAWE config.yml result in any changes to this?

**Comment by mibby** [comment]
Doesn't ``allowed-plugins`` only make it so notifications of slowdown caused by that plugin never gets a message sent to console?

CraftBook works fine with FAWE 1.12 and vanilla WorldEdit 1.13, so I assume it is something in FAWE 1.13 causing the breakage.

**Comment by IronApollo** [comment]
Iirc adding a plugin to the allowed-plugins list suppresses extent creation warnings from wrapping extents created by the plugin in question. That would not have any effect on the issue described.

Can you describe what exactly the gate mechanic is? I have no experienced with that plugin.

**Comment by mibby** [comment]
http://wiki.sk89q.com/wiki/CraftBook/Gates

**Comment by mibby** [comment]
Gate mechanic is still broken with FAWE as of commit https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/commit/85bfd16d7c16613e79f49d3acaa17007c09baa3e (3/25 compile).
CraftBook dev 4491

**Comment by mibby** [comment]
Weekly re-test. Still broke as of commit https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/commit/8afd96f03b3ea797208a681672d620eb62e1b9b7.

**Comment by AeSix** [comment]
The devs will post an update when they attempt to resolve this issue.  Empire is back working on FAWE, and there may be some work towards this in the near future.  It may also be resolved once FAWE is API compatible with WorldEdit 7.  As far as my understanding, they're going targeting WE7 beta 5 right now until beta 6 is out.  It's not easy for them to keep targeting snapshots.  

FAWE is not production ready, and you should use WorldEdit 7 until FAWE is ready.

**Comment by mibby** [comment]
@boy0001 For reference, getting this error now at startup with FastAsyncWorldEdit-Breaking dev 2.

```
[00:34:13] [Server thread/WARN]: [CraftBook] Failed to load mechanic: Gate
java.lang.NoClassDefFoundError: com/sk89q/worldedit/world/block/FuzzyBlockState
	at java.lang.Class.getDeclaredConstructors0(Native Method) ~[?:?]
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:3138) ~[?:?]
	at java.lang.Class.getConstructor0(Class.java:3343) ~[?:?]
	at java.lang.Class.newInstance(Class.java:556) ~[?:?]
	at com.sk89q.craftbook.bukkit.CraftBookPlugin.setupCraftBook(CraftBookPlugin.java:563) ~[?:?]
	at com.sk89q.craftbook.bukkit.CraftBookPlugin.onEnable(CraftBookPlugin.java:451) ~[?:?]
	at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:265) ~[patched_1.13.2.jar:git-Paper-603]
	at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:334) ~[patched_1.13.2.jar:git-Paper-603]
	at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:412) ~[patched_1.13.2.jar:git-Paper-603]
	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.enablePlugin(CraftServer.java:443) ~[patched_1.13.2.jar:git-Paper-603]
	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.enablePlugins(CraftServer.java:357) ~[patched_1.13.2.jar:git-Paper-603]
	at net.minecraft.server.v1_13_R2.MinecraftServer.l(MinecraftServer.java:608) ~[patched_1.13.2.jar:git-Paper-603]
	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:570) ~[patched_1.13.2.jar:git-Paper-603]
	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:431) ~[patched_1.13.2.jar:git-Paper-603]
	at net.minecraft.server.v1_13_R2.DedicatedServer.init(DedicatedServer.java:316) ~[patched_1.13.2.jar:git-Paper-603]
	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:787) ~[patched_1.13.2.jar:git-Paper-603]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.ClassNotFoundException: com.sk89q.worldedit.world.block.FuzzyBlockState
	at java.net.URLClassLoader.findClass(URLClassLoader.java:471) ~[?:?]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:140) ~[patched_1.13.2.jar:git-Paper-603]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:86) ~[patched_1.13.2.jar:git-Paper-603]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:588) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:521) ~[?:?]
	... 17 more
```

**Comment by boy0001** [comment]
Ah, it uses fuzzy blocks.

**Comment by Tsoccerguy3** [comment]
Just tested all our Gates on our server , they are all not working and same error , as of the latest commit today

**Comment by Tsoccerguy3** [comment]
Is there a fix thatt anyone has found ?

**Comment by AeSix** [comment]
according to bstats, less than 10% of FAWE installations are using CraftBook.  That's assuming ALL craftbook installs are also using FAWE.  Taking a guess at only half of the CraftBook installs are using FAWE, that's roughly 330 servers minimum utilizing both.  At most, there's 678 utilizing both. In comparison to the nearly 8700 installs of FAWE.  

My point, is that yes, there's a lot of CraftBook servers out there, but it's a small percentage of issues when compared to all of the other things which still need to be resolved.  The devs still have plenty to do with base Minecraft mechanics before worrying about a somewhat niche plugin. 

I am sure that the devs will eventually fix the issues, if possible.  But please understand that CraftBook compatibility is a pretty low priority in comparison to all of the other issues they're facing.  And please remember that none of the devs are full time paid employees focusing solely on FAWE - they are all volunteers writing and fixing as they have time and interest. 

I'm also certain that the devs will note in the changelogs, or close the ticket once there is an update worth your time to test.

Additionally, perhaps this is not actually an issue which FAWE must resolve, but should be resolved with CraftBook fixes. Maybe *politely* ask @me4502  if he is aware of the issue, and could possibly take a look at it, as he knows the CraftBook code.  If nothing else, he may be able to work with the FAWE devs, or even supply a PR to FAWE which can be used to correct the issue. 

If you wish to continue to test CB/FAWE builds and update those test results, it would be very helpful to all to know some basic information:
Server implimentation:
Server build/version:
FAWE build:
CraftBook build:
And if you are running other plugins or not. 

I suggest testing with Spigot (and Paper as well if you wish) with as close to default configs as you can use with your setup, using the last build available for 1.13.2 (I wouldn't even touch 1.14 any time soon for testing) And making sure you're using the release builds and dev builds of the plugins.  This should be done with every new build (release or dev/breaking/experimental) build of FAWE and CB.  Or wait until the devs say they've attempted to resolve the issue and use builds containing that fix. 

At least then bumps to this issue will have useful, relevant data.

**Comment by me4502** [comment]
Do note that CraftBook’s bStats setup was broken for a large amount of time, so there’s only like 5 builds where it counts.

If it works with WorldEdit, but not FAWE, it’s 100% a FAWE issue and should be fixed in FAWE.

**Comment by Tsoccerguy3** [comment]
> 
> 
> Do note that CraftBook’s bStats setup was broken for a large amount of time, so there’s only like 5 builds where it counts.
> 
> If it works with WorldEdit, but not FAWE, it’s 100% a FAWE issue and should be fixed in FAWE.

I see that its a fuzzy blocks issue in FAWE , Do you have a fix that you can push to FAWE ?
Thanks for taking the time to look at this .

**Comment by Tofuus** [comment]
Jesse removed it in this commit 33f5322fda3afd3c9a3ca07028511a3772093db6. Presumably to replace it with his own solution.

**Comment by me4502** [comment]
That's considered an issue with FAWE then. If they're breaking the API contracts of the WorldEdit API, then it's not filling the dependency requirement of WorldEdit.

If it's not going to implement the API, it should stop pretending to be WorldEdit. This is 99% of the reason I preferred it using patching.

**Comment by AeSix** [comment]
@me4502 I appologize if I sounded like I was blaming you or CraftBook.  I don't know the code of any of the projects and was merely stating that it was a possibility.  I didn't realized CraftBook had an issue with bstats either.  However, my main object was to get @Tsoccerguy3 to understand that there is a lot at play here, a lot of issues in FAWE, a lot of code to check/write/fix... and not a lot of time for the devs to focus on what (at least at the time) seemed to be a niche combination of plugins, and that they would more than likely address the issue in time.  

Thank you for replying and setting things straight.  I also agree that FAWE seems to have strayed considerably far from WorldEdit, but I know that the devs want to maintain drop-in replacement with WorldEdit.  With the devs being more or less unfamiliar with the code, Jesse's absence and the massive amounts of changes, I think it's a bit premature to drop the idea of FAWE being a drop-in replacement.  Though, I've not be involved as heavily lately, I still believe this all to be accurate.  

@Tsoccerguy3 is there a build including the changes from https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/commit/33f5322fda3afd3c9a3ca07028511a3772093db6 for which you can test with CraftBook?  It would help to know if it fixes the issue.  I would test it myself, however I have never once used CraftBook, and would take a week to learn enough to test it properly.

**Comment by Tsoccerguy3** [comment]
> 
> 
> That's considered an issue with FAWE then. If they're breaking the API contracts of the WorldEdit API, then it's not filling the dependency requirement of WorldEdit.
> 
> If it's not going to implement the API, it should stop pretending to be WorldEdit. This is 99% of the reason I preferred it using patching.

Hi Matt , Just to clarify , the last version that Craftbook [Gates] was working , was a snapshot I compiled for 1.12.2 . It worked with World Edit and FAWE in Spigot 1.12.2 .

Currently there has never been a Craftbook build that [Gate] has worked in 1.13.2 using World Edit or FAWE .  The ID's are messed up and incomplete .

Testing , If you test to see if [Gate] works , you will find that it does not . 
Try the final build of Spigot 1.13.2 with the latest compile of World Edit 7 and the latest compile of Craftbook . You will see that the various materials can not be identified ; example ... , a fence has 2 states a}. a Post and b}. a Fence .

I have tried loading the sign with various materials , selecting and pasting materials into the [Gate] frame and manually placing materials in to the frame . None work .

The config file is out of date with new 1.13.2 names , not really a problem if the names were internally corrected to be in sync with the names in the config .

Thanks for helping out

**Comment by Tsoccerguy3** [comment]
> 
> 
> @me4502 I appologize if I sounded like I was blaming you or CraftBook. I don't know the code of any of the projects and was merely stating that it was a possibility. I didn't realized CraftBook had an issue with bstats either. However, my main object was to get @Tsoccerguy3 to understand that there is a lot at play here, a lot of issues in FAWE, a lot of code to check/write/fix... and not a lot of time for the devs to focus on what (at least at the time) seemed to be a niche combination of plugins, and that they would more than likely address the issue in time.
> 
> Thank you for replying and setting things straight. I also agree that FAWE seems to have strayed considerably far from WorldEdit, but I know that the devs want to maintain drop-in replacement with WorldEdit. With the devs being more or less unfamiliar with the code, Jesse's absence and the massive amounts of changes, I think it's a bit premature to drop the idea of FAWE being a drop-in replacement. Though, I've not be involved as heavily lately, I still believe this all to be accurate.
> 
> @Tsoccerguy3 is there a build including the changes from [33f5322](https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/commit/33f5322fda3afd3c9a3ca07028511a3772093db6) for which you can test with CraftBook? It would help to know if it fixes the issue. I would test it myself, however I have never once used CraftBook, and would take a week to learn enough to test it properly.

Read the comment left for Matt , Tested for Mibby to try to narrow down what's going on . I concluded , feature incomplete or WIP , I how ever tested his issue , did you ? I read through your post and you stated the obvious but nothing helpful . The obvious is , World Edit 7 still has issues with the states of fences and other ID's

**Comment by Tsoccerguy3** [comment]
Trying the various ways to add the material to the [Gate] frame .

![2019-04-29_10 01 51](https://user-images.githubusercontent.com/22108822/56976785-baa60700-6b41-11e9-8d6a-69bc33d3851d.png)
![2019-04-29_10 02 11](https://user-images.githubusercontent.com/22108822/56976786-baa60700-6b41-11e9-9224-67fccc963018.png)
![2019-04-29_10 02 33](https://user-images.githubusercontent.com/22108822/56976787-bb3e9d80-6b41-11e9-8031-da2b9155cc09.png)
![2019-04-29_10 08 37](https://user-images.githubusercontent.com/22108822/56976788-bb3e9d80-6b41-11e9-92f5-be2d894814a7.png)
![2019-04-29_10 09 08](https://user-images.githubusercontent.com/22108822/56976789-bb3e9d80-6b41-11e9-8afa-493fe280b36c.png)
![2019-04-29_10 09 36](https://user-images.githubusercontent.com/22108822/56976790-bb3e9d80-6b41-11e9-8dd1-3bc2ec8c7826.png)
![2019-04-29_10 10 03](https://user-images.githubusercontent.com/22108822/56976791-bb3e9d80-6b41-11e9-8ee6-4a03b289d9e2.png)
![2019-04-29_10 10 20](https://user-images.githubusercontent.com/22108822/56976792-bb3e9d80-6b41-11e9-9fca-1f5efead8fbb.png)

**Comment by me4502** [comment]
Gates work perfectly fine with WorldEdit, as does the rest of CraftBook. If it’s not for you, you’re either not using the latest versions or you’ve change the configuration and broken it.

**Comment by Tsoccerguy3** [comment]
Another Gate anomaly , because the gate frame has a decorative edge on the outside , the gate gets made from intersecting fence state . This is from todays compile of worldedit and craftbook. 

![2019-05-01_20 31 42](https://user-images.githubusercontent.com/22108822/57052108-32b02200-6c53-11e9-87bd-d5d18c5b6be4.png)
![2019-05-01_20 32 08](https://user-images.githubusercontent.com/22108822/57052109-32b02200-6c53-11e9-88e0-7e5a5221b0f7.png)

**Comment by Tsoccerguy3** [comment]
> 
> 
> Ah, it uses fuzzy blocks.

When you do fininish updating the issue with fuzzy blocks , as you can see there is also a problem with the different states of Fence , glass panes and iron bars. the gates worked ok in 1.12.2

**Comment by AeSix** [comment]
@Tsoccerguy3 I just built a server to test this; I installed:
WorldEdit 7 build 4238
CraftBook 3 build 4492
Both plugins have default configurations, with the exception of adding "- Gate" to the enabled mechanics list in CraftBook's config. 

I then went and learned how to make a gate because I have never used CraftBook before. 

I successfully built and operated a basic gate.  There is no issues with the fence state, there is no issue with the operation, it just worked like I expect a gate to work. All but the top row of fence disappear upon "opening" the gate (as is explained as proper behavior), and are recreated upon "closing" the gate. 

The ONLY exception to the above is when using a decorative edging on the sides/top of the gate as in your last example.  Yes, the fences have dangling connectors.  This is because of the overlapping stone at the top of the gate, which the top row of fence are connected to.  I assume CraftBook simply stacks the top most row of fence down when "closing", and thus duplicating the gate's state exactly as it is in the top remaining row.   This can be overcome by setting the gate material, by hand with WorldEdit to oak_fence, or by allowing for either an air block on both face sides of the gate's top row, or using blocks which fence cannot connect to. 
 
I personally like the spiked gate feeling this "bug" gives, and as it is using standard WE mechanics to duplicate the fence downward, it's not something I see needs to be resolved within WorldEdit, but possibly withing CraftBook as a "ensure smooth gates" config option which would then toggle code that would go through and verify each fence section and correct it's state upon "closing" - which will slow down the gate closing.  IF it's even considered a bug, but FAWE is not the place to determine that.

I have also confirmed that no recent build of FAWE works with the gates, including 105, which was built very recently, and at this time is the newest build for 1.13.

To summaraize:
CraftBook with WE7 works with expected behavior. 
CraftBook with FAWE does NOT work, and says so very clearly in the console. 
The devs are well aware of the issue.  I am sure when one of them finally tackles fuzzyblocks and is confident in the code's functional operation towards this end, he will make it known.  
Additionally, I have asked if anyone is currently working on fuzzyblock code for FAWE. ~~~I will update this post with the response when I receive one.~~~ UPDATE: NMF seems to believe this should be working.

In the mean time, there is no point in testing every new build if there is no mention of fuzzyblocks / craftbook / gates / fences.

If you ABSOLUTELY MUST have functioning CraftBook, use WorldEdit 7 beta5, or one of the dev builds, specifically 4238 as it is known to work.

**Comment by Tsoccerguy3** [comment]
> 
> 
> Gates work perfectly fine with WorldEdit, as does the rest of CraftBook. If it’s not for you, you’re either not using the latest versions or you’ve change the configuration and broken it.

I tested your latest Craftbook build 4500 with Fawe 's breaking build (the fuzzy block commit ) , I can now get the [Gate] working 1 time by setting , //set 85 or 101, the old ID's for Oak Fence and Iron Fence , but that just sets the post variation of the fence or glass pane. The [Gate] will open 1 time up and 1 time down The material is now conected to each other but not to the [Gate] frame as in the pictures above. On subsequent toggles , it can not find gate
 . Setting the material Manually and Toggle gate , it says the [Gate] is toggled , but it is not
Stocking the sign shows an increase in material loaded to the sign , but does nothing else

These are my findings ,

**Comment by AeSix** [comment]
Excellent!  Not that it's still not working, but good descriptions for Jesse to be able to figure it out better. 

Are there any errors in the console/log?  Those would help as well!

**Comment by Tsoccerguy3** [comment]
Output From Console , no errors , just WARN
[CraftBook] Enabling CraftBook v3.10-SNAPSHOT;4500-4cecc6e
[15:58:44] [Server thread/INFO]: [CraftBook] Loading persistent data from YAML!
[15:58:44] [Server thread/INFO]: [CraftBook] Successfully added 1 CommandItems!
[15:58:44] [Server thread/INFO]: [CraftBook] Registered 3 custom recipes!
[15:58:44] [Server thread/INFO]: [CraftBook] 1 cauldron recipe(s) loaded
[15:58:44] [Server thread/INFO]: [CraftBook] Enumerating chunks for self-triggered components...
[15:58:44] [Server thread/INFO]: [CraftBook] 1875 chunk(s) for 3 world(s) processed (37ms elapsed)
[15:58:44] [Server thread/WARN]: [CraftBook] ====================================================
[15:58:44] [Server thread/WARN]: [CraftBook]  CraftBook works better if you use Paper 
[15:58:44] [Server thread/WARN]: [CraftBook]  as your server software. 
[15:58:44] [Server thread/WARN]: [CraftBook]   
[15:58:44] [Server thread/WARN]: [CraftBook]  Paper offers significant performance improvements,
[15:58:44] [Server thread/WARN]: [CraftBook]  bug fixes, security enhancements and optional
[15:58:44] [Server thread/WARN]: [CraftBook]  features for server owners to enhance their server.
[15:58:44] [Server thread/WARN]: [CraftBook]   
[15:58:44] [Server thread/WARN]: [CraftBook]  Paper includes Timings v2, which is significantly
[15:58:44] [Server thread/WARN]: [CraftBook]  better at diagnosing lag problems over v1.
[15:58:44] [Server thread/WARN]: [CraftBook]   
[15:58:44] [Server thread/WARN]: [CraftBook]  All of your plugins should still work, and the
[15:58:44] [Server thread/WARN]: [CraftBook]  Paper community will gladly help you fix any issues.
[15:58:44] [Server thread/WARN]: [CraftBook]   
[15:58:44] [Server thread/WARN]: [CraftBook]  Join the Paper Community @ https://papermc.io
[15:58:44] [Server thread/WARN]: [CraftBook] ====================================================

**Comment by Tsoccerguy3** [comment]
Tried with Paper624 . No Errors or Warnings on anything

**Comment by AeSix** [comment]
does "or anything" mean "or working gates" or does it mean that it's working properly with that combination?

**Comment by Tsoccerguy3** [comment]
> 
> 
> does "or anything" mean "or working gates" or does it mean that it's working properly with that combination?
I was surprised after updating all plugins to the latest available , the console was quiet of errors and warnings , plugins preferring Paper over Spigot .

I mean , the console error is gone . I guess the final issue is with the gate recognizing the new names and there states . The gate works 1 time the material changes . I think the problem is very close to being resolved at this point

**Comment by Tsoccerguy3** [comment]
As a control , I tested

WorldEdit build 4253
Craftbook build 4500

Gates are working ,

**Comment by Tsoccerguy3** [comment]
I have found the reason why the [Gate] will only work 1 time up and 1 time down .

https://minecraft.gamepedia.com/Block_states   , look under Fence

oak fence , old ID = 85 or new ID  , mincraft:oak_fence   is just the unconnected oak post . player is unable to place this fence state manually because you will always get a connected fence state . When you set the [Gate] material as in //set minecraft:oakfence you will get a [Gate] made up of just oak fence posts as in the picture above . This will satisfy the declared material in the mechanics config file , so the [Gate] will work 1 time up and 1 time down . When the fence is down the connected state of the fence is no longer minecraft:oak_fence , but is in another state such as example ,  minecraft:oak_fence,north,true . 

Since the gates work in all Worldedit versions with craftbook , the fix may be to just look to see how Worldedit handles the different states of fences etc... vs. FAWE

**Comment by Tsoccerguy3** [comment]
![2019-05-09_23 15 58](https://user-images.githubusercontent.com/22108822/57500280-e0a17900-72b0-11e9-953b-51f87d997430.png)
![2019-05-09_23 16 06](https://user-images.githubusercontent.com/22108822/57500281-e0a17900-72b0-11e9-9d29-db2d0d7740f4.png)

Constructing a [Gate] with spaces between the posts , works perfectly opening and closing .

**Comment by Tsoccerguy3** [comment]
![2019-05-09_23 17 11](https://user-images.githubusercontent.com/22108822/57500372-2d854f80-72b1-11e9-9619-8c1f64ff84f3.png)
![2019-05-09_23 17 33](https://user-images.githubusercontent.com/22108822/57500373-2e1de600-72b1-11e9-87c5-1b04853a9712.png)

constructing a [Gate] with connected fence , fails to find gate or operate at all

**Comment by Tsoccerguy3** [comment]
Hi Matt , had a question about what to put in the mechanisms.yml to create this Fence state on purpose , player really like the 3D effect  that the [Gate] has 

Gate:
        allow-redstone: true
        limit-columns: true
        max-columns: 14
        blocks:
        - minecraft:acacia_fence
        - minecraft:birch_fence
        - minecraft:jungle_fence
        - minecraft:oak_fence
        - minecraft:spruce_fence
        - minecraft:dark_oak_fence
        - minecraft:nether_brick_fence
        - minecraft:iron_bars
        - minecraft:glass_pane
        enforce-type: true
        max-column-height: 20
        gate-search-radius: 7

https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/issues/51#issuecomment-488523650

**Comment by AeSix** [comment]
@Tsoccerguy3 what happens if you replace "minecraft:oak_fence" with "oak_fence"?  Is there any changes?

**Comment by Tsoccerguy3** [comment]
> 
> 
> @Tsoccerguy3 what happens if you replace "minecraft:oak_fence" with "oak_fence"? Is there any changes?

it makes no difference , failed to find gate

**Comment by Tsoccerguy3** [comment]
> 
> 
> That's considered an issue with FAWE then. If they're breaking the API contracts of the WorldEdit API, then it's not filling the dependency requirement of WorldEdit.
> 
> If it's not going to implement the API, it should stop pretending to be WorldEdit. This is 99% of the reason I preferred it using patching.

So if FAWE updates from WorldEdit Beta 5 dependencies , to the Release candidate 1 this will fix the block names issues , ? this seems to be the issue now

**Comment by me4502** [comment]
Nothing related to block names has changed in a very long time, it’s not a matter of FAWE not being updated. It’s just broken parts of the API, specifically fuzzy matching.

**Comment by Tsoccerguy3** [comment]
> 
> 
> Nothing related to block names has changed in a very long time, it’s not a matter of FAWE not being updated. It’s just broken parts of the API, specifically fuzzy matching.

So a fuzzy state refers to , example: when a fence post can have a connector in each of the directions and at all 4 sides ?

When this is fixed in FAWE how can I set a fence state in the config file .

See https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/issues/51#issuecomment-491283055

**Comment by bloggy** [comment]
I have the same issue. Does anybody know when and if this can be fixed?

**Comment by Tsoccerguy3** [comment]
> I have the same issue. Does anybody know when and if this can be fixed?
The problem is fuzzy blocks and it is partially implemented . It is still being worked on in breaking dev.
some blocks like fence change there state depending on the block next to them .
there is no time to say when it will be fixed in FAWE . It does work with Worldedit 7 RC2

**Comment by dartzdeRS** [comment]
dev fix this plz?

**Comment by Tsoccerguy3** [comment]
> 
> 
> Nothing related to block names has changed in a very long time, it’s not a matter of FAWE not being updated. It’s just broken parts of the API, specifically fuzzy matching.

Matt , could you help Jesse by patching the broken parts of fuzzy matching .  ? Maybe its time to update the World Edit API from Beta 5 to the RC 2 or is it better to wait for a final ?  It may also fix a issue with version 1.12.2 schematics are pasted into 1.13.2 and some fence , stairs , redstone , are not connected together or in the wrong state

**Comment by bloggy** [comment]
Do they even read the github issues ? :D

**Comment by me4502** [comment]
The gate config allows specifying anything.

And I don't have time to fix plugins that break WorldEdit.

**Comment by MattBDev** [comment]
I'm doing the best I can in my limited time. I am actively reading this thread. My priority was on bringing FAWE up to speed on the text component changes. I'll reevaluate what needs to be done first.

**Comment by TheoRetisch1** [comment]
Confirmed for the latest version. The gates can be registered, but when the gates are triggered the error "Failed to find a gate" appears! No error in the console. 

FAWE: 1.13-113
craftbook: 3.10-SNAPSHOT;4500-4cecc6e

**Comment by Tsoccerguy3** [comment]
As of today FAWE only recognizes fuzzy blocks default names . when the new names and their states are defined , it will fix this issue and a lot more issues mentioned in other issues

**Comment by TheoRetisch1** [comment]
@Tsoccerguy3 could you please tell me with which craftbook and FAWE version it works for you? I am now on FAWE 1.13-117 and have reinstalled everything. The problem with the Craftbook Gates still exists.

**Comment by Tsoccerguy3** [comment]
> 
> 
> @Tsoccerguy3 could you please tell me with which craftbook and FAWE version it works for you? I am now on FAWE 1.13-117 and have reinstalled everything. The problem with the Craftbook Gates still exists.

At this time only this configuration works , there has to be a air space between the base material to maintain default state of example minecraft:oak_fence,[false] . or by old naming standard ID 85 which is a oak post not connected to anything. Name matching does not work because the code from Worldedit is missing from FAWE

https://user-images.githubusercontent.com/22108822/57500280-e0a17900-72b0-11e9-953b-51f87d997430.png

**Comment by Tsoccerguy3** [comment]
Name | Default value | Allowed values | Description
-- | -- | -- | --
north | eastnorthsouthwest | For an open gate, the direction the gates swing open.For a closed gate, the direction the player was facing when the gate was placed, or the last direction the gates have swung.
false | falsetrue | If true, the gate is lowered by three pixels, to accommodate attaching more cleanly with walls.
false | falsetrue | If true, the gate is opened.
false | falsetrue | If true, the gate is receiving redstone power.

**Comment by TheoRetisch1** [comment]
Thank you for the detailed explanation :). I hope that the problem will be fixed soon.

**Comment by Tsoccerguy3** [comment]
> 
> 
> The gate config allows specifying anything.
> 
> And I don't have time to fix plugins that break WorldEdit.

Matt can you give an example of the material name for oak_fence with forks on all 4 sides [ north , south , east , west ] to put in Craftbooks allowed material section of the config . I'm guess that it would be similar for glass panes and iron bars as well.

**Comment by AeSix** [comment]
@Tsoccerguy3
If you want me4502's help with Craftbook, please use the resources available from the Enginhub/Craftbook community.  Make sure you are using WorldEdit, not FAWE when requesting assistance from them.  You can use what you learn to apply to FAWE.   This is not the correct place to ask him for assistance, as he has made it clear, and it's been proven, that Craftbook works properly with WorldEdit, and that any issues arising from CB's use with FAWE is not his concern.  

It's neither fair nor right to ask developers of other projects for assistance through a project they are not developers on.  me4502 can work with the FAWE developers behind the scene if he so chooses, and which he has done so in the past. 

When FAWE and CraftBook work correctly you can update this ticket (and close), or when you find new issues with compatibility, those should be posted as new tickets to FAWE's issue tracker. 

Additionally, It would be nice if you were to refer to MattBDev or me4502 by those names, respectively, as "Matt" is ambiguous and leads to confusion.

**Comment by Tsoccerguy3** [comment]
> 
> 
> @Tsoccerguy3
> If you want me4502's help with Craftbook, please use the resources available from the Enginhub/Craftbook community. Make sure you are using WorldEdit, not FAWE when requesting assistance from them. You can use what you learn to apply to FAWE. This is not the correct place to ask him for assistance, as he has made it clear, and it's been proven, that Craftbook works properly with WorldEdit, and that any issues arising from CB's use with FAWE is not his concern.
> 
> It's neither fair nor right to ask developers of other projects for assistance through a project they are not developers on. me4502 can work with the FAWE developers behind the scene if he so chooses, and which he has done so in the past.
> 
> When FAWE and CraftBook work correctly you can update this ticket (and close), or when you find new issues with compatibility, those should be posted as new tickets to FAWE's issue tracker.
> 
> Additionally, It would be nice if you were to refer to MattBDev or me4502 by those names, respectively, as "Matt" is ambiguous and leads to confusion.

Your right , but at this point I'm not even sure if the issue exists anymore . I can not seem to find what the proper name to use to test FAWE . I've searched the internet to get an example of the name to use , with no luck . Asking for an example , people dance around the question but no answer .  Does anyone know what the name is?


AeSix , can you tell me what the name is? if i had the name i could generate the block in FAWE with a //set command ,

**Comment by Tsoccerguy3** [comment]
MIbby should close this issue , because it is just a symptom of a bigger issue . Then a new ticket opened dealing with the new name structure .

it affects schematics
redstone 
connecting glass panes , iron bars , coblestone fence 
masking
filters
etc ...

need FAWE code to deal with

get Block_Name
get Block_State  ,  its either true of false

the new name   minecraft:Block_Name[north=Block_State,south=Block_State,east=Block_State,west=Block_State]

**Comment by Tsoccerguy3** [comment]
> I'm doing the best I can in my limited time to keep up on the issues. The other Matt (me4502) doesn't have to do anything for this project if he doesn't want to. I am actively reading this thread. My priority was on bringing FAWE up to speed on the text component changes. I'll reevaluate what needs to be done first.

MatBDev , I can see that you have committed thousands of lines of code in the last few months , Thank You by the way . Since you know this issue better than anyone , could you not just enter the status on this issue and close it ? This post is all about a symptom that effects the [Gate} in CraftBook , but this should be closed and all the issues with , changing block states , opened in a new ticket . I see at least 6 Tickets open , that are the same thing . The gate now does work as long as the declared  material stays in its default state  .   
Thank You

**Comment by Tsoccerguy3** [comment]
looking though code , I see that it;s unfinished worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditPlugin.java .  Starting line 168 to end , Marked as TODO ,  So the missing code that stops the gate from working is here and a work in progress

**Comment by Tsoccerguy3** [comment]
> 
> 
> I'm doing the best I can in my limited time to keep up on the issues. The other Matt (me4502) doesn't have to do anything for this project if he doesn't want to. I am actively reading this thread. My priority was on bringing FAWE up to speed on the text component changes. I'll reevaluate what needs to be done first.

MattDev are you any closer to fixing this bug ? In this case it affects Craftbook  and some other issues in other posts

**Comment by Tsoccerguy3** [comment]
Status update , I see Tofuus did a fix for this issue . Although it does not work . The sign is stocked with Air and all powered redstone on the map becomes surrounded by lamps , lmao

**Comment by Tsoccerguy3** [comment]
Status update

FAWE 1.13 breaking build 146 , breaks craftbook totally . Probably because you can no longer use any version of worldguard with it . I'm guessing MattDev is in the process of updating FAWE to use the latest Worldgard and Worldedit API's .

**Comment by MattBDev** [comment]
@Tsoccerguy3 please reply with an error log using the latest version of craftbook. I am overhauling everything at the moment.

**Comment by Tsoccerguy3** [comment]
No Problem , I'll get that to you

**Comment by Tsoccerguy3** [comment]
I used worldguard 7.0 final and the latest code for craftbook
Here is our server log https://pastebin.com/3V3zrfes

-----------------------------------------------------------------------------------------

using the last worldguard build 1847 that worked with FAWE worldguard does load but no Craftbook functionality
https://pastebin.com/11xwiDU5

**Comment by Tsoccerguy3** [comment]
So the issue with Craftbook Gate is fixed . I see you are in the process of updating to Worldguard and Worldedit 7 final API .

Great work guys

**Comment by Tsoccerguy3** [comment]
Someone should close this ticket , as fixed . post is getting to long

**Comment by Tsoccerguy3** [comment]
> 
> 
> @Tsoccerguy3 please reply with an error log using the latest version of craftbook. I am overhauling everything at the moment.

Hi MattDev , I did not want to open another ticket involving caftbook and the blockstate issue .
I just checked Fawe breaking 1.13 build 170 , someone implemented a wildcard fix to find the the gate fences , but without any filters relating to the declared material in craftbooks mechanics.yml  . The first block that is found before the fence material is air , so a gate filled with and sign stocked with air. Any way .Gate does not work for that reason. This was Toofus's idea to use the wildcard and he had the same problem that the first wildcard block found was air . This info might help to fix . Thanks
 
![2019-07-10_21 48 39](https://user-images.githubusercontent.com/22108822/61020585-46bb8280-a36c-11e9-9c4d-9cc79cd5711b.png)
![2019-07-10_21 48 43](https://user-images.githubusercontent.com/22108822/61020586-47541900-a36c-11e9-8e28-867f06fd7ad2.png)
![2019-07-10_21 48 50](https://user-images.githubusercontent.com/22108822/61020587-47541900-a36c-11e9-8da9-32bc34a2dc68.png)
![2019-07-10_21 49 21](https://user-images.githubusercontent.com/22108822/61020588-47541900-a36c-11e9-89d6-d5f8271d45bc.png)
![2019-07-10_23 30 04](https://user-images.githubusercontent.com/22108822/61020589-47541900-a36c-11e9-81db-0198e9d2e8fe.png)

**Comment by Tsoccerguy3** [comment]
I revisited testing Caftbook gates with FAWE merge 151 and it is still broke if anyone wants to test

**Comment by Tsoccerguy3** [comment]
Tested in paper 1.15.1 build 25 . FastAsyncWorldEdit-1.15-5.jar the craftbook gate mechanic is still broken . This feature does work in Worldedit 7.1

**Comment by Tsoccerguy3** [comment]
This is Fixed in 1.15 build 8 someone should close

**Comment by N0tMyFaultOG** [comment]
Fixed in https://github.com/IntellectualSites/FastAsyncWorldEdit-1.13/commit/c2cc4f5aeb9e65fd32ab18eac0a126addca57001
